File: ServerControl/webview2-client/src/shared/pal-config.ts
import type { SettingMetadata, SettingType } from './types.ts'

export interface PalDocument {
  prefix: string
  suffix: string
  entries: Map<string, string>
  order: string[]
}

function isEscaped(text: string, index: number): boolean {
  let backslashes = 0
  for (let behind = index - 1; behind >= 0 && text[behind] === '\\'; behind -= 1) backslashes += 1
  return backslashes % 2 !== 0
}

/** Splits only top-level commas, preserving quoted strings and tuple contents byte-for-byte. */
export function splitPalOptionSegments(text: string): string[] {
  const segments: string[] = []
  let segmentStart = 0
  let depth = 0
  let inQuotes = false

  for (let index = 0; index < text.length; index += 1) {
    const character = text[index]
    if (character === '"' && !isEscaped(text, index)) {
      inQuotes = !inQuotes
      continue
    }
    if (inQuotes) continue
    if (character === '(') depth += 1
    else if (character === ')') depth -= 1
    else if (character === ',' && depth === 0) {
      segments.push(text.slice(segmentStart, index))
      segmentStart = index + 1
    }
  }

  segments.push(text.slice(segmentStart))
  return segments
}

/** Parses OptionSettings while retaining all surrounding and unknown INI content. */
export function readPalWorldSettings(content: string): PalDocument {
  const marker = 'OptionSettings=('
  const markerIndex = content.indexOf(marker)
  if (markerIndex < 0) throw new Error('OptionSettings section was not found.')

  const innerStart = markerIndex + marker.length
  let depth = 1
  let inQuotes = false
  let innerEnd = -1

  for (let index = innerStart; index < content.length; index += 1) {
    const character = content[index]
    if (character === '"' && !isEscaped(content, index)) {
      inQuotes = !inQuotes
      continue
    }
    if (inQuotes) continue
    if (character === '(') depth += 1
    else if (character === ')') {
      depth -= 1
      if (depth === 0) {
        innerEnd = index
        break
      }
    }
  }

  if (innerEnd < 0) throw new Error('OptionSettings closing parenthesis was not found.')

  const entries = new Map<string, string>()
  const order: string[] = []
  for (const segment of splitPalOptionSegments(content.slice(innerStart, innerEnd))) {
    const equalsIndex = segment.indexOf('=')
    if (equalsIndex <= 0) throw new Error(`Invalid OptionSettings segment: ${segment}`)
    const name = segment.slice(0, equalsIndex).trim()
    if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(name)) throw new Error(`Invalid setting name: ${name}`)
    if (entries.has(name)) throw new Error(`Duplicate setting found: ${name}`)
    entries.set(name, segment.slice(equalsIndex + 1))
    order.push(name)
  }

  return {
    prefix: content.slice(0, innerStart),
    suffix: content.slice(innerEnd),
    entries,
    order,
  }
}

/** Applies known changes without normalizing or dropping any untouched setting. */
export function writePalWorldSettings(document: PalDocument, changes: ReadonlyMap<string, string>): string {
  for (const name of changes.keys()) {
    if (!document.entries.has(name)) throw new Error(`Cannot update missing setting: ${name}`)
  }
  const segments = document.order.map((name) => `${name}=${changes.get(name) ?? document.entries.get(name) ?? ''}`)
  return document.prefix + segments.join(',') + document.suffix
}

export function unquotePalValue(rawValue: string): string {
  if (rawValue.length >= 2 && rawValue.startsWith('"') && rawValue.endsWith('"')) {
    return rawValue.slice(1, -1).replace(/\\"/g, '"').replace(/\\\\/g, '\\')
  }
  return rawValue
}

export function decodePalValue(rawValue: string, type: SettingType): string {
  return type === 'string' || type === 'secret' ? unquotePalValue(rawValue) : rawValue
}

export function encodePalValue(value: string, metadata: SettingMetadata): string {
  switch (metadata.type) {
    case 'boolean':
      if (!/^(true|false)$/i.test(value)) throw new Error(`${metadata.name} must be True or False.`)
      return value.toLowerCase() === 'true' ? 'True' : 'False'
    case 'integer': {
      if (!/^-?\d+$/.test(value.trim())) throw new Error(`${metadata.name} must be a whole number.`)
      const parsed = Number(value)
      if (!Number.isSafeInteger(parsed)) throw new Error(`${metadata.name} is outside the supported whole-number range.`)
      validateRange(parsed, metadata)
      return String(parsed)
    }
    case 'number': {
      const parsed = Number(value)
      if (!Number.isFinite(parsed) || value.trim() === '') throw new Error(`${metadata.name} must be a finite number.`)
      validateRange(parsed, metadata)
      return trimNumber(parsed)
    }
    case 'enum': {
      const option = metadata.options?.find((candidate) => candidate.toLowerCase() === value.toLowerCase())
      if (!option) throw new Error(`${metadata.name} must be one of: ${(metadata.options ?? []).join(', ')}.`)
      return option
    }
    case 'string':
    case 'secret':
      return `"${value.replace(/\\/g, '\\\\').replace(/"/g, '\\"')}"`
    case 'raw':
      return value
  }
}

function validateRange(value: number, metadata: SettingMetadata): void {
  if (metadata.min !== undefined && value < metadata.min) throw new Error(`${metadata.name} must be at least ${metadata.min}.`)
  if (metadata.max !== undefined && value > metadata.max) throw new Error(`${metadata.name} must be at most ${metadata.max}.`)
}

export function trimNumber(value: number): string {
  return value.toFixed(6).replace(/\.?0+$/, '')
}

export function displayValue(value: string, metadata: SettingMetadata): string {
  if (metadata.type !== 'number') return value
  const parsed = Number(value)
  return Number.isFinite(parsed) ? trimNumber(parsed) : value
}

File: ServerControl/webview2-client/host/PalConfig.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;

namespace PalworldServerControl.Host;

internal sealed record PalDocument(string Prefix, string Suffix, Dictionary<string, string> Entries, List<string> Order);

internal static class PalConfig
{
    internal static IReadOnlyList<string> SplitOptionSegments(string text)
    {
        var result = new List<string>();
        var start = 0;
        var depth = 0;
        var quoted = false;
        for (var index = 0; index < text.Length; index++)
        {
            var character = text[index];
            if (character == '"')
            {
                var slashes = 0;
                for (var behind = index - 1; behind >= 0 && text[behind] == '\\'; behind--) slashes++;
                if (slashes % 2 == 0) quoted = !quoted;
                continue;
            }
            if (quoted) continue;
            if (character == '(') depth++;
            else if (character == ')') depth--;
            else if (character == ',' && depth == 0)
            {
                result.Add(text[start..index]);
                start = index + 1;
            }
        }
        result.Add(text[start..]);
        return result;
    }

    internal static PalDocument Read(string content)
    {
        const string marker = "OptionSettings=(";
        var markerIndex = content.IndexOf(marker, StringComparison.Ordinal);
        if (markerIndex < 0) throw new InvalidOperationException("OptionSettings section was not found.");
        var innerStart = markerIndex + marker.Length;
        var depth = 1;
        var quoted = false;
        var innerEnd = -1;
        for (var index = innerStart; index < content.Length; index++)
        {
            var character = content[index];
            if (character == '"')
            {
                var slashes = 0;
                for (var behind = index - 1; behind >= 0 && content[behind] == '\\'; behind--) slashes++;
                if (slashes % 2 == 0) quoted = !quoted;
                continue;
            }
            if (quoted) continue;
            if (character == '(') depth++;
            else if (character == ')' && --depth == 0) { innerEnd = index; break; }
        }
        if (innerEnd < 0) throw new InvalidOperationException("OptionSettings closing parenthesis was not found.");

        var entries = new Dictionary<string, string>(StringComparer.Ordinal);
        var order = new List<string>();
        foreach (var segment in SplitOptionSegments(content[innerStart..innerEnd]))
        {
            var equals = segment.IndexOf('=');
            if (equals <= 0) throw new InvalidOperationException($"Invalid OptionSettings segment: {segment}");
            var name = segment[..equals].Trim();
            if (name.Length == 0 || !(char.IsAsciiLetter(name[0]) || name[0] == '_') ||
                name.Any(character => !(char.IsAsciiLetterOrDigit(character) || character == '_')))
                throw new InvalidOperationException($"Invalid setting name: {name}");
            if (!entries.TryAdd(name, segment[(equals + 1)..]))
                throw new InvalidOperationException($"Duplicate setting found: {name}");
            order.Add(name);
        }
        return new PalDocument(content[..innerStart], content[innerEnd..], entries, order);
    }

    internal static string Write(PalDocument document, IReadOnlyDictionary<string, string> changes)
    {
        foreach (var name in changes.Keys)
            if (!document.Entries.ContainsKey(name)) throw new InvalidOperationException($"Cannot update missing setting: {name}");
        return document.Prefix + string.Join(',', document.Order.Select(name =>
            $"{name}={(changes.TryGetValue(name, out var value) ? value : document.Entries[name])}")) + document.Suffix;
    }

    internal static string Decode(string rawValue, string type) =>
        type is "string" or "secret" ? Unquote(rawValue) : rawValue;

    internal static string Encode(string value, SettingMetadata metadata)
    {
        switch (metadata.Type)
        {
            case "boolean":
                if (!bool.TryParse(value, out var boolean)) throw new InvalidOperationException($"{metadata.Name} must be True or False.");
                return boolean ? "True" : "False";
            case "integer":
                if (!long.TryParse(value.Trim(), NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out var integer))
                    throw new InvalidOperationException($"{metadata.Name} must be a whole number.");
                ValidateRange(integer, metadata);
                return integer.ToString(CultureInfo.InvariantCulture);
            case "number":
                if (string.IsNullOrWhiteSpace(value) || !double.TryParse(value, NumberStyles.Float, CultureInfo.InvariantCulture, out var number) || !double.IsFinite(number))
                    throw new InvalidOperationException($"{metadata.Name} must be a finite number.");
                ValidateRange(number, metadata);
                return TrimNumber(number);
            case "enum":
                var option = metadata.Options?.FirstOrDefault(candidate => string.Equals(candidate, value, StringComparison.OrdinalIgnoreCase));
                return option ?? throw new InvalidOperationException($"{metadata.Name} must be one of: {string.Join(", ", metadata.Options ?? [])}.");
            case "string":
            case "secret":
                return $"\"{value.Replace("\\", "\\\\", StringComparison.Ordinal).Replace("\"", "\\\"", StringComparison.Ordinal)}\"";
            case "raw": return value;
            default: throw new InvalidOperationException($"Unsupported setting type: {metadata.Type}");
        }
    }

    internal static string Display(string value, SettingMetadata metadata) =>
        metadata.Type == "number" && double.TryParse(value, NumberStyles.Float, CultureInfo.InvariantCulture, out var number)
            ? TrimNumber(number) : value;

    internal static string Sha256(byte[] content) => Convert.ToHexString(SHA256.HashData(content));
    internal static string Sha256(string content) => Sha256(Encoding.UTF8.GetBytes(content));

    private static string Unquote(string value) => value.Length >= 2 && value[0] == '"' && value[^1] == '"'
        ? value[1..^1].Replace("\\\"", "\"", StringComparison.Ordinal).Replace("\\\\", "\\", StringComparison.Ordinal)
        : value;

    private static string TrimNumber(double value) => value.ToString("0.######", CultureInfo.InvariantCulture);

    private static void ValidateRange(double value, SettingMetadata metadata)
    {
        if (metadata.Min is not null && value < metadata.Min) throw new InvalidOperationException($"{metadata.Name} must be at least {metadata.Min}.");
        if (metadata.Max is not null && value > metadata.Max) throw new InvalidOperationException($"{metadata.Name} must be at most {metadata.Max}.");
    }
}
